OneSignalSDK/onesignal/src/main/java/com/onesignal/GenerateNotification.java

```diff
@@ -114,12 +112,20 @@ static void fromJsonPayload(final OSNotificationGenerationJob notifJob) {
          return;
 
       // New thread so that any Network usage is off the main thread before showing the notification
-      new Thread(new Runnable() {
+      Thread thread = new Thread(new Runnable() {
          @Override
          public void run() {
             showNotification(notifJob);
          }
-      }, OS_SHOW_NOTIFICATION_THREAD).start();
+      }, OS_SHOW_NOTIFICATION_THREAD);
+      thread.start();
+
+      // Stop all code until thread is done doing work to show notification
+      try {
+         thread.join();
```

Comment:
Would easier to explain over a call if you have time.

Basically the join method will await the current thread until the new thread is done with its work. It does not immediately join actually.

These [docs](https://docs.oracle.com/javase/tutorial/essential/concurrency/join.html) explain it well as well as the diagram below
![Thread Join in Java](https://user-images.githubusercontent.com/24537305/89454193-5b647880-d72e-11ea-927f-54d6100034a5.png)

Reply:
But where is the 2nd thread we are kicking off and where are we joining it?
Still looking through the PR so that may uncover this question.

Reply:
Ahh I understand, so there are two places we call `GenerateNotification.fromJsonPayload` and both routes come from this method `NotificationBundleProcessor.processJobForDisplay`

One is at this line
https://github.com/OneSignal/OneSignal-Android-SDK/blob/76b94f52ea4aabba6b580710dc35a692042780ea/OneSignalSDK/onesignal/src/main/java/com/onesignal/NotificationBundleProcessor.java#L127

Other is at this line
https://github.com/OneSignal/OneSignal-Android-SDK/blob/76b94f52ea4aabba6b580710dc35a692042780ea/OneSignalSDK/onesignal/src/main/java/com/onesignal/OSNotificationGenerationJob.java#L301

Reply:
Also that diagram is only showing you how the `Thread.join` method works, we only kick off Thread 1 technically and its only for the `showNotification` method

Reply:
Update based on our discussion.
Adding the thread here prevents throwing an exception of downloading on the main thread, however this can create an ANR if downloading takes 5+ sec

Reply:
Another update is we are not calling `showNotification` from Main Thread and therefore the `new Thread` is not necessary. Good talk though and after meeting with Josh to discuss this further we decided we should add a `RuntimeException` check right before the `showNotification` method [PR for Main Thread RuntimeException check](https://github.com/OneSignal/OneSignal-Android-SDK/pull/1110)